distributed/comm/ucx.py

```diff
@@ -100,6 +138,72 @@ def device_array(n):
                     "In order to send/recv CUDA arrays, Numba or RMM is required"
                 )
 
+    # Find the functions `device_concat` and `device_split`
+    try:
+        import cupy
+
+        def device_concat(arys):
+            with cupy.cuda.using_allocator(rmm.rmm_cupy_allocator):
+                arys = [cupy.asarray(e).view("u1") for e in arys]
+                result = cupy.concatenate(arys, axis=None)
+                result_buffer = result.data.mem._owner
+                return result_buffer
+
+        def device_split(ary, indices):
+            with cupy.cuda.using_allocator(rmm.rmm_cupy_allocator):
+                ary = cupy.asarray(ary).view("u1")
+                ary_split = cupy.split(ary, indices)
+                results = []
+                result_buffers = []
+                for e in ary_split:
+                    e2 = cupy.empty_like(e)
+                    cupy.copyto(e2, e)
+                    results.append(e2)
+                    result_buffers.append(e2.data.mem._owner)
+                cupy.cuda.stream.get_current_stream().synchronize()
```

Comment:
Note that we synchronize once to ensure all of the copies did complete. We do this to make sure the data is valid before the `DeviceBuffer` it came from is freed.

Reply:
Once the `DeviceBuffer` goes out of scope I believe the actual freeing of memory is enqueued onto the stream, so you should be fine due to stream ordering. @harrism @jrhemstad is that correct from the perspective of RMM?

If this is just a normal cuda memory allocation that will call `cudaFree` then that also gets enqueued onto the stream so you should be safe in general without this synchronize I believe.

Reply:
Cool I went ahead and dropped it. If we need it back, it is easy to revert.

Reply:
You need to make sure that the buffer's memory was never used on a different stream than the one it will be deleted on without synchronizing the streams before deleting.

[Answer to Keith's question: It's not strictly to correct to say the freeing of memory is enqueued onto the stream. Just that an allocation on the same stream will be allowed to use the freed block in stream order (e.g. safely). An allocation on a different stream will only use that block after the streams are synchronized. ]